redash/tasks/queries/maintenance.py

```diff
@@ -27,85 +27,86 @@ def empty_schedules():
-                            "Skipping refresh of {} because a related dropdown "
-                            "query ({}) is unattached to any datasource."
-                        ).format(query.id, e.query_id)
-                        track_failure(query, error)
-                        continue
-
-                enqueue_query(
-                    query_text,
-                    query.data_source,
-                    query.user_id,
-                    scheduled_query=query,
-                    metadata={"Query ID": query.id, "Username": "Scheduled"},
+
+def _apply_default_parameters(queries):
+    for query in queries:
+        parameters = {p["name"]: p.get("value") for p in query.parameters}
+        if any(parameters):
+            try:
+                query.query_text = query.parameterized.apply(parameters).query
+            except InvalidParameterError as e:
+                error = u"Skipping refresh of {} because of invalid parameters: {}".format(
+                    query.id, str(e)
                 )
+                track_failure(query, error)
+                continue
+            except QueryDetachedFromDataSourceError as e:
+                error = (
+                    "Skipping refresh of {} because a related dropdown "
+                    "query ({}) is unattached to any datasource."
+                ).format(query.id, e.query_id)
+                track_failure(query, error)
+                continue
 
-                query_ids.append(query.id)
-                outdated_queries_count += 1
+        yield query
 
-    statsd_client.gauge("manager.outdated_queries", outdated_queries_count)
 
-    logger.info(
-        "Done refreshing queries. Found %d outdated queries: %s"
-        % (outdated_queries_count, query_ids)
-    )
+def _enqueue_queries(queries):
+    for query in queries:
+        try:
+            enqueue_query(
+                query.query_text,
+                query.data_source,
+                query.user_id,
+                scheduled_query=query,
+                metadata={"Query ID": query.id, "Username": "Scheduled"},
+            )
+            
+            yield query
+        except Exception as e:
+            logging.info("Could not enqueue query %d due to %s", query.id, repr(e))
 
-    status = redis_connection.hgetall("redash:status")
-    now = time.time()
 
-    redis_connection.hmset(
-        "redash:status",
-        {
-            "outdated_queries_count": outdated_queries_count,
-            "last_refresh_at": now,
-            "query_ids": json_dumps(query_ids),
-        },
-    )
+def refresh_queries():
+    logger.info("Refreshing queries...")
 
-    statsd_client.gauge(
-        "manager.seconds_since_refresh", now - float(status.get("last_refresh_at", now))
-    )
+    outdated = models.Query.outdated_queries()
+    refreshable = _skip_unrefreshable_queries(outdated)
+    queries = _apply_default_parameters(refreshable)
+    enqueued = list(_enqueue_queries(queries))
```

Comment:
Wouldn't it be easier to understand, more performant (less loops) and simpler if it was something like:

```python
for query in models.Query.outdated_queries():
    # logic from _skip_unrefreshable_queries
    if not should_refresh_query(query):
        continue
    try:
        enqueue_query(
            # logic from _apply_default_parameters
            apply_default_parameters(query.query_text),
            query.data_source,
            query.user_id,
            scheduled_query=query,
            metadata={"Query ID": query.id, "Username": "Scheduled"},
      )
    except Exception as e:
        logging.info("Could not enqueue query %d due to %s", query.id, repr(e))
```

Reply:
Also, similar comment about exception logging as with the one in `outdated_queries`. Except that here we should only report to Sentry, but not unschedule.

Reply:
Simplicity / understandability is a matter of preference. Personally I thought of the journey from `outdated_queries` to `enqueued` as a set of transformations and filters, so I took a FP-based approach. I've switched to your suggestion in ef2eb39.

Regarding more performant - not sure, those were generator loops and I think there's a built-in optimization there.